### ISSUE TITLE
False positive `unnecessary-direct-lambda-call` / `C3002` with assignment expression

### ISSUE THREAD

**Issue by ColeFrench** [pylint-dev/pylint#9294]
### Bug description

The walrus strikes again! Here is a minimum reproducible example (`mre.py`) that false positives `unnecessary-direct-lambda-call`:

```python
# pylint: disable=missing-docstring,too-few-public-methods
class C:
    assert (lambda: all((x := object()) is x for _ in range(1)))()
```

This is a false positive due to [a sneaky clause in PEP 572 – Assignment Expressions](https://peps.python.org/pep-0572/#:~:text=A%20further%20exception%20applies%20when%20an%20assignment%20expression%20occurs%20in%20a%20comprehension%20whose%20containing%20scope%20is%20a%20class%20scope.) that prohibits assignment expressions in a comprehension within a class scope.

Above, the comprehension is a generator expression within the lambda's scope, so the code compiles. But if we were to remove the lambda as the warning suggests, then the expression would lie in the class' scope, so the code would not compile (`dnc.py`):

```python
# pylint: disable=missing-docstring,too-few-public-methods
class C:
    assert all((x := object()) is x for _ in range(1))
```

```
  File "dnc.py", line 3
    assert all((x := object()) is x for _ in range(1))
               ^
SyntaxError: assignment expression within a comprehension cannot be used in a class body
```

The example itself demonstrates an important use case that is difficult or maybe even impossible to work around: referencing the exact same object twice in a comprehension.

### Command used

```shell
pylint mre.py
```

### Pylint output

```
************* Module mre
mre.py:3:11: C3002: Lambda expression called directly. Execute the expression inline instead. (unnecessary-direct-lambda-call)

------------------------------------------------------------------
Your code has been rated at 5.00/10

```

### Expected behavior

```

--------------------------------------------------------------------
Your code has been rated at 10.00/10

```

### Pylint version

```
pylint 3.0.2
astroid 3.0.1
Python 3.8.18 | packaged by conda-forge | (default, Oct 10 2023, 15:44:36)
[GCC 12.3.0]
```

### OS / Environment

Debian GNU/Linux 12 (bookworm)

**Comment by chiruu12** [comment]
Taking this one, PR shortly.

Still reproduces on `4.1.0-dev0` / astroid `4.3.0`, and following the message is a `SyntaxError`:

```
assignment expression within a comprehension cannot be used in a class body
```

All four comprehension forms hit it, and only in a class body:

```
genexp / listcomp / setcomp / dictcomp with a walrus, in a class body   SyntaxError
walrus without a comprehension, in a class body                        fine
comprehension without a walrus, in a class body                        fine
any of them inside a function or at module level                       fine
```

So the guard is narrow: the call's frame is a `ClassDef`, and the lambda body holds a comprehension
with a `NamedExpr` that would land in that class body once the lambda is gone. A comprehension
nested inside a further lambda keeps a scope of its own after inlining, so that case still reports.
